src/advanced/timestamping-tutorial.md

```diff
@@ -0,0 +1,929 @@
+    }
+```
+
+As we intend for the endpoints to be accessinble to external users, we declare
+them as within the `public_scope` of API endpoints. We create the following
+endpoints:
+
+- For the `handle_timestamp` method we declare the `v1/timestamps/value` route,
+  which will be used for checking whether a certain timestamping transactions
+  exists in the blockchain.
+- For the `handle_timestamp_proof` method we declare the `v1/timestamps/proof`
+  route, which will be used for getting proofs of correctness of certain
+  timestamping transactions.
+- For the `handle_post_transaction` method we declare the `v1/timestamps` route,
+  which will be used for sending timestamping transactions to the blockchain
+  network.
+
+## Define Service
+
+Service is a group of templated transactions (we have defined them before).
+It has a name and a unique ID to determine the service inside the blockchain. We
+define our new service in [src/lib.rs][src/lib.rs]:
+
+```rust
+#[derive(Debug, Default)]
+pub struct Service;
+
+impl Service {
+    pub fn new() -> Self {
+        Service
+    }
+}
+```
+To turn `Service` into a blockchain service, we need to implement the Exonum
+`Service` trait for it.
+
+!!!tip
+   Read more on how to turn a type into a blockchain service in the
+   [Interface with Exonum Framework](../architecture/services.md#interface-with-exonum-framework)
+   section.
+
+```rust
+impl blockchain::Service for Service {
+    fn service_id(&self) -> u16 {
+        TIMESTAMPING_SERVICE
+    }
+
+    fn service_name(&self) -> &'static str {
+        SERVICE_NAME
+    }
+
+    fn state_hash(&self, view: &dyn Snapshot) -> Vec<Hash> {
+        let schema = Schema::new(view);
+        schema.state_hash()
+    }
+
+    fn tx_from_raw(&self, raw: RawTransaction) -> Result<Box<dyn Transaction>, StreamStructError> {
+        let tx = TimeTransactions::tx_from_raw(raw)?;
+        Ok(tx.into())
+    }
+
+    fn wire_api(&self, builder: &mut ServiceApiBuilder) {
+        PublicApi::wire(builder);
+    }
+}
+```
+The code above declares a number of methods:
+- `service_id` - returns the ID of the service
+- `service_name` - returns the name of the service
+- `state_hash` - calculates the hash of
+  [the blockchain state](../glossary.md#blockchain-state). The method
+  should return [a vector of hashes](../architecture/services.md#state-hash)
+  of the [Merkelized service tables](../glossary.md#merkelized-indices).
+- `tx_from_raw` - deserializes transactions processed by the service.
+   If the incoming transaction is built successfully, we put it into a `Box<_>`.
+- `wire_api` - connects the API endpoints of the service to the API of the
+  blockchain.
+
+## Create Service Instance
+
+We define the `ServiceFactory` structure in [src/lib.rs][src/lib.rs]. This
+structure builds an instance of our service.
+
+```rust
+#[derive(Debug, Clone, Copy)]
+pub struct ServiceFactory;
+```
+The `ServiceFactory` structure compiles an instance of the Timestamping service
```

Comment:
I don't like the word `compiles` when there is nothing about actual compilation. That is the case. ServiceFactory is used to make certain service applicable for `NodeBuilder`, nothing more